modules/host/contractmanager/sectorremoval.go

```diff
@@ -0,0 +1,232 @@
+package contractmanager
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"go.sia.tech/siad/crypto"
+	"go.sia.tech/siad/modules"
+)
+
+const removalEntrySize = 20
+
+type (
+	removalEntry struct {
+		ID sectorID
+		// Count refers to the number of references that should be removed from
+		// the contract manager.
+		Count uint64
+		// Offset refers to the position in the persistence file of the entry
+		Offset int64
+	}
+
+	sectorRemovalMap struct {
+		mu          sync.Mutex
+		sectors     map[sectorID]removalEntry
+		entries     int64
+		removalFile *os.File
+
+		cm  *ContractManager
+		wal *writeAheadLog
+	}
+)
+
+// writes the removal entry to disk
+func (srm *sectorRemovalMap) writeRemovalEntry(id sectorID, count uint64, offset int64) error {
+	buf := make([]byte, removalEntrySize)
+	copy(buf[:], id[:])
+	binary.LittleEndian.PutUint64(buf[12:], count)
+	_, err := srm.removalFile.WriteAt(buf, offset)
+	return err
+}
+
+// managedLoad loads the removal queue from disk. Entries with a count of zero
+// and read errors are ignored. The file is compacted while loading.
+func (srm *sectorRemovalMap) managedLoad(path string) (err error) {
+	srm.mu.Lock()
+	defer srm.mu.Unlock()
+
+	srm.removalFile, err = os.OpenFile(path, os.O_RDWR|os.O_CREATE, modules.DefaultFilePerm)
+	if err != nil {
+		return errors.AddContext(err, "unable to open sector removal queue file")
+	}
+	br := bufio.NewReader(srm.removalFile)
+	buf := make([]byte, removalEntrySize)
+	// read each entry from disk. The persistence file is compacted while
+	// reading.
+	for i := int64(0); ; i++ {
+		if _, err := io.ReadFull(br, buf); err == io.EOF || err == io.ErrUnexpectedEOF {
+			break
+		} else if err != nil {
+			return errors.AddContext(err, "error reading sector removal file")
+		}
+
+		entry := removalEntry{
+			Count:  binary.LittleEndian.Uint64(buf[12:]),
+			Offset: srm.entries * removalEntrySize,
+		}
+		copy(entry.ID[:], buf)
+		// skip entries that are no longer needed.
+		if entry.Count == 0 {
+			continue
+		}
+
+		srm.sectors[entry.ID] = entry
+		// if the number of read entries is greater than the number of used
+		// entries overwrite the unused entries.
+		if srm.entries != i {
+			if err := srm.writeRemovalEntry(entry.ID, entry.Count, entry.Offset); err != nil {
```

Comment:
is the idea here that all of the entries "shift down" to fill in the gap? That would be O(n) in the number of entries, right? Should we be concerned about that?

Reply:
Yes, all entries will shift down to fill gaps; it makes tracking the file offset of entries relatively simple. I don't think we need to be concerned since it only happens when the file is loaded. We could batch the writes to reduce syscalls or add a temp file+rename at the end of the load.